Orange/widgets/utils/owlearnerwidget.py

```diff
@@ -1,106 +1,53 @@
+from copy import deepcopy
 import numpy as np
+
 from AnyQt.QtCore import QTimer, Qt
 
-from Orange.classification.base_classification import LearnerClassification
 from Orange.data import Table
-from Orange.modelling import Fitter
+from Orange.modelling import Fitter, Learner, Model
 from Orange.preprocess.preprocess import Preprocess
-from Orange.regression.base_regression import LearnerRegression
 from Orange.widgets import gui
-from Orange.widgets import widget
 from Orange.widgets.settings import Setting
+from Orange.widgets.utils import getmembers
+from Orange.widgets.utils.signals import Output, Input
 from Orange.widgets.utils.sql import check_sql_input
 from Orange.widgets.widget import OWWidget, WidgetMetaClass, Msg
 
 
-class DefaultWidgetChannelsMetaClass(WidgetMetaClass):
-    """Metaclass that adds default inputs and outputs objects.
-    """
-
-    REQUIRED_ATTRIBUTES = []
-
-    def __new__(mcls, name, bases, attrib):
-        # check whether it is abstract class
-        if attrib.get('name', False):
-            # Ensure all needed attributes are present
-            if not all(attr in attrib for attr in mcls.REQUIRED_ATTRIBUTES):
-                raise AttributeError(
-                    "'{}' must have '{}' attributes"
-                    .format(name, "', '".join(mcls.REQUIRED_ATTRIBUTES)))
-
-            attrib['outputs'] = mcls.update_channel(
-                mcls.default_outputs(attrib),
-                attrib.get('outputs', [])
-            )
+class OWBaseLearnerMeta(WidgetMetaClass):
+    """ Meta class for learner widgets
 
-            attrib['inputs'] = mcls.update_channel(
-                mcls.default_inputs(attrib),
-                attrib.get('inputs', [])
-            )
+    OWBaseLearner declares two outputs, learner and model with
+    generic type (Learner and Model).
 
-            mcls.add_extra_attributes(name, attrib)
-
-        return super().__new__(mcls, name, bases, attrib)
-
-    @classmethod
-    def default_inputs(cls, attrib):
-        return []
-
-    @classmethod
-    def default_outputs(cls, attrib):
-        return []
-
-    @classmethod
-    def update_channel(cls, channel, items):
-        item_names = set(item[0] if isinstance(item, tuple) else item.name
-                         for item in channel)
-
-        for item in items:
-            if item[0] not in item_names:
-                channel.append(item)
-
-        return channel
-
-    @classmethod
-    def add_extra_attributes(cls, name, attrib):
-        return attrib
-
-
-class OWBaseLearnerMeta(DefaultWidgetChannelsMetaClass):
-    """Metaclass that adds default inputs (table, preprocess) and
-    outputs (learner, model) for learner widgets.
+    This metaclass ensures that each of the subclasses gets
+    its own Outputs class with output that match the corresponding
+    learner.
     """
+    def __new__(cls, name, bases, attributes):
```

Comment:
As far as I know, first argument to methods should be either self or cls if it is a class method (which __new__ is).

https://www.python.org/dev/peps/pep-0008/#id46